File: python/infinicore/ops/hbmv.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def hbmv(
    alpha: Tensor,
    a: Tensor,
    x: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    uplo: int = 0,
    k: int,
):
    _infinicore.hbmv_(
        alpha._underlying,
        a._underlying,
        x._underlying,
        beta._underlying,
        out._underlying,
        uplo,
        k,
    )

    return out

File: src/infinicore/ops/hbmv/hbmv.cc
#include "infinicore/ops/hbmv.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(Hbmv);

Hbmv::Hbmv(const Tensor &alpha, const Tensor &a, const Tensor &x, const Tensor &beta, Tensor y, int uplo, size_t k) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(alpha, a, x, beta, y);
    INFINICORE_GRAPH_OP_DISPATCH(y->device().getType(), alpha, a, x, beta, y, uplo, k);
}

void Hbmv::execute(const Tensor &alpha, const Tensor &a, const Tensor &x, const Tensor &beta, Tensor y, int uplo, size_t k) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(Hbmv, alpha, a, x, beta, y, uplo, k);
}

void hbmv_(const Tensor &alpha, const Tensor &a, const Tensor &x, const Tensor &beta, Tensor y, int uplo, size_t k) {
    Hbmv::execute(alpha, a, x, beta, y, uplo, k);
}

} // namespace infinicore::op
